cns/multitenantcontroller/multitenantoperator/multitenantcrdreconciler.go

```diff
@@ -0,0 +1,156 @@
+package multitenantoperator
+
+import (
+	"context"
+	"net"
+	"strings"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/Azure/azure-container-networking/cns/cnsclient"
+	"github.com/Azure/azure-container-networking/cns/logger"
+	ncapi "github.com/Azure/azure-container-networking/networkcontainer/api/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	NCStateInitialized = "Initialized" // NC has been initialized by DNC.
+	NCStateSucceeded   = "Succeeded"   // NC has been persisted by CNS
+	NCStateTerminated  = "Terminated"  // NC has been cleaned up from CNS
+)
+
+// multiTenantCrdReconciler reconciles multi-tenant network containers.
+type multiTenantCrdReconciler struct {
+	KubeClient client.Client
+	NodeName   string
+	CNSClient  cnsclient.APIClient
+}
+
+// Reconcile is called on multi-tenant CRD status changes.
+func (r *multiTenantCrdReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	ctx := context.Background()
+	var nc ncapi.NetworkContainer
+
+	if err := r.KubeClient.Get(ctx, request.NamespacedName, &nc); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+
+		logger.Errorf("Failed to fetch network container: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	if !nc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// Do nothing if the NC has already in Terminated state.
+		if nc.Status.State == NCStateTerminated {
+			return ctrl.Result{}, nil
+		}
+
+		// Remove the deleted network container from CNS.
+		err := r.CNSClient.DeleteNC(cns.DeleteNetworkContainerRequest{
+			NetworkContainerid: nc.Spec.UUID,
+		})
+		if err != nil {
+			logger.Errorf("Failed to delete NC %s from CNS: %v", nc.Spec.UUID, err)
+			return ctrl.Result{}, err
+		}
+
+		// Update NC state to Terminated.
+		nc.Status.State = NCStateTerminated
+		if err := r.KubeClient.Status().Update(ctx, &nc); err != nil {
+			logger.Errorf("Failed to update network container state for %s: %v", nc.Spec.UUID, err)
+			return ctrl.Result{}, err
+		}
+
+		logger.Printf("NC has been terminated for %s", nc.Spec.UUID)
+		return ctrl.Result{}, nil
+	}
+
+	// Do nothing if the network container hasn't been initialized yet from control plane.
+	if nc.Status.State != NCStateInitialized {
+		return ctrl.Result{}, nil
```

Comment:
Should we return error here so the request will be requeued?

Reply:
No, it should wait for the next update when the state changes.